graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/processor/EntityMappingModelBuilder.java

```diff
@@ -0,0 +1,296 @@
+        udtCqlSchema = null;
+        break;
+      case UDT:
+        if (regularColumns.isEmpty()) {
+          invalidMapping("%s must have at least one field", graphqlName);
+          return Optional.empty();
+        }
+        tableCqlSchema = null;
+        udtCqlSchema = buildCqlUdt(context.getKeyspace().name(), cqlName, regularColumns);
+        break;
+      default:
+        throw new AssertionError("Unexpected target " + target);
+    }
+
+    return isFederated(partitionKey, clusteringColumns, target)
+        .map(
+            isFederated ->
+                new EntityMappingModel(
+                    graphqlName,
+                    context.getKeyspace().name(),
+                    cqlName,
+                    target,
+                    partitionKey,
+                    clusteringColumns,
+                    regularColumns,
+                    tableCqlSchema,
+                    udtCqlSchema,
+                    isFederated,
+                    inputTypeName));
+  }
+
+  private String providedCqlNameOrDefault(Optional<Directive> cqlEntityDirective) {
+    return cqlEntityDirective
+        .flatMap(d -> DirectiveHelper.getStringArgument(d, "name", context))
+        .orElse(graphqlName);
+  }
+
+  private EntityMappingModel.Target providedTargetOrDefault(
+      Optional<Directive> cqlEntityDirective) {
+    return cqlEntityDirective
+        .flatMap(
+            d ->
+                DirectiveHelper.getEnumArgument(
+                    d, "target", EntityMappingModel.Target.class, context))
+        .orElse(EntityMappingModel.Target.TABLE);
+  }
+
+  private String providedInputNameOrDefault(Directive cqlInputDirective) {
+    Optional<String> maybeName =
+        DirectiveHelper.getStringArgument(cqlInputDirective, "name", context);
+    if (maybeName.isPresent()) {
+      return maybeName.get();
+    } else {
+      info(
+          "%1$s: using '%1$sInput' as the input type name since @cql_input doesn't "
+              + "have an argument",
+          graphqlName);
+      return graphqlName + "Input";
+    }
+  }
+
+  private boolean hasPartitionKey(
+      List<FieldMappingModel> partitionKey, List<FieldMappingModel> regularColumns) {
+    if (!partitionKey.isEmpty()) {
+      return true;
+    }
+    FieldMappingModel firstField = regularColumns.get(0);
+    if (TypeHelper.mapsToUuid(firstField.getGraphqlType())) {
+      info(
+          "%s: using %s as the partition key, "
+              + "because it has type %s and no other fields are annotated",
+          graphqlName, firstField.getGraphqlName(), firstField.getGraphqlType());
+      partitionKey.add(firstField.asPartitionKey());
+      regularColumns.remove(firstField);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Check that if the @key directive is present, it matches the CQL primary key.
+   *
+   * @return empty if the directive is invalid (an error has already been logged).
+   */
+  private Optional<Boolean> isFederated(
```

Comment:
This signature is... not great. What we would really need here is some sort of `Try` type, not `Optional`. But let's not go there.

But really it makes me consider replacing the whole `Optional` thing by a checked `SkipException`.